pkg/querier/querier.go

```diff
@@ -181,38 +196,106 @@ func (q querier) Select(sp *storage.SelectParams, matchers ...*labels.Matcher) (
+	// time series from store and time series from ingesters).
+	return q.mergeSeriesSets(result), nil, nil
 }
 
 // LabelsValue implements storage.Querier.
 func (q querier) LabelValues(name string) ([]string, storage.Warnings, error) {
-	lv, err := q.distributor.LabelValuesForLabelName(q.ctx, model.LabelName(name))
-	return lv, nil, err
+	return q.metadataQuerier.LabelValues(name)
 }
 
 func (q querier) LabelNames() ([]string, storage.Warnings, error) {
-	ln, err := q.distributor.LabelNames(q.ctx)
-	return ln, nil, err
+	return q.metadataQuerier.LabelNames()
 }
 
-func (q querier) metadataQuery(matchers ...*labels.Matcher) (storage.SeriesSet, storage.Warnings, error) {
-	ms, err := q.distributor.MetricsForLabelMatchers(q.ctx, model.Time(q.mint), model.Time(q.maxt), matchers...)
-	if err != nil {
-		return nil, nil, err
+func (querier) Close() error {
+	return nil
+}
+
+func (q querier) mergeSeriesSets(sets []storage.SeriesSet) storage.SeriesSet {
+	// Here we deal with sets that are based on chunks and build single set from them.
+	// Remaining sets are merged with chunks-based one using storage.NewMergeSeriesSet
+
+	otherSets := []storage.SeriesSet(nil)
+	chunks := []chunk.Chunk(nil)
+
+	for _, set := range sets {
+		if !set.Next() {
+			// nothing in this set
+			continue
+		}
+
+		s := set.At()
+		if sc, ok := s.(SeriesWithChunks); ok {
+			chunks = append(chunks, sc.Chunks()...)
+
+			// iterate over remaining series in this set, and store chunks
+			// Here we assume that all remaining series in the set are also backed-up by chunks.
+			// If not, there will be panics.
+			for set.Next() {
+				s = set.At()
+				chunks = append(chunks, s.(SeriesWithChunks).Chunks()...)
+			}
+		} else {
+			// We already called set.Next() once, but we want to return same result from At() also
+			// to the query engine.
+			otherSets = append(otherSets, &seriesSetWithFirstSeries{set: set, firstSeries: s})
+		}
+	}
+
+	if len(chunks) == 0 {
+		return storage.NewMergeSeriesSet(otherSets, nil)
+	}
+
+	// partitionChunks returns set with sorted series, so it can be used by NewMergeSeriesSet
+	chunksSet := partitionChunks(chunks, q.mint, q.maxt, q.chunkIterFn)
+
+	if len(otherSets) == 0 {
+		return chunksSet
 	}
-	return metricsToSeriesSet(ms), nil, nil
+
+	otherSets = append(otherSets, chunksSet)
+	return storage.NewMergeSeriesSet(otherSets, nil)
 }
 
-func (querier) Close() error {
-	return nil
+// This series set ignores first 'Next' call and simply returns cached result
+// to avoid doing the work required to compute it twice.
+type seriesSetWithFirstSeries struct {
+	firstNextCalled bool
+	firstSeries     storage.Series
+	set             storage.SeriesSet
+}
+
+func (pss *seriesSetWithFirstSeries) Next() bool {
+	if pss.firstNextCalled {
+		pss.firstSeries = nil
+		return pss.set.Next()
+	}
+	pss.firstNextCalled = true
+	return true
+}
+
+func (pss *seriesSetWithFirstSeries) At() storage.Series {
+	if pss.firstSeries != nil {
```

Comment:
Despite correct, it looks weird to me that this `if` is different then the one in the `Next()`. Wouldn't be more clear if we check `pss.firstNextCalled` here too? If `firstSeries` is `nil` we return `nil`. Am I wrong?

Reply:
In situation where first `set.Next()` returns true, but at the same time Err() also returns non-nil, this would hide that error. I think I better add a comment to the code. WDYT?

Reply:
> In situation where first set.Next() returns true, but at the same time Err() also returns non-nil, this would hide that error. 

Why would the error hidden? I can't see a difference in error handling either we check `pss.firstSeries != nil` or `!pss.firstNextCalled`. Could you help me understand the scenario you're describing, please?

Reply:
Scenario I'm talking about is this:

`mergeSeriesSets` calls `set.Next()` and then `set.At()` to get first series. `mergeSeriesSets` doesn't call `set.Err()` at all, but maybe this call would already return some error at this point if it was actually called. This is the error that would be hidden if we did the check in `seriesSetWithFirstSeries.Err()` method. (It wouldn't be lost, as it would be returned on next `Err()` call).

But looking at the actually usage, it seems that `Err()` is only checked when `Next()` returns false, so your suggestion would actually work.

I think there is a bug in `mergeSeriesSets` – we shouldn't just ignore empty sets, but return their errors if there are any. I'll fix that.